cpp/src/arrow/dataset/expression_benchmark.cc

```diff
@@ -0,0 +1,105 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <iostream>
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/cast.h"
+#include "arrow/dataset/expression.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+#include "arrow/util/range.h"
+
+namespace arrow {
+namespace dataset {
+
+static std::shared_ptr<Partitioning> MakePartitioning() {
+  auto schema = arrow::schema({
+      arrow::field("a", arrow::dictionary(arrow::int32(), arrow::int64())),
+      arrow::field("b", arrow::dictionary(arrow::int32(), arrow::int64())),
+  });
+  std::shared_ptr<Array> a_dict, b_dict;
+  auto values = internal::Iota<int64_t>(100);
+  ArrayFromVector<Int64Type>(arrow::int64(), values, &a_dict);
+  ArrayFromVector<Int64Type>(arrow::int64(), values, &b_dict);
+  ArrayVector dictionaries = {a_dict, b_dict};
+  return std::make_shared<HivePartitioning>(schema, dictionaries);
+}
+
+static Expression GetPartitionExpression(const std::string& path) {
+  static std::shared_ptr<Partitioning> partitioning = MakePartitioning();
+  ASSIGN_OR_ABORT(auto expr, partitioning->Parse(path));
+  return expr;
+}
```

Comment:
To maintain parity with the python case which motivated this benchmark (and to keep the boilerplate of constructing a `Partitioning` minimal), let's use a `PartitioningFactory` here:

```suggestion
static Expression GetPartitionExpression(const std::string& path, bool infer_dictionary) {
  auto options = HivePartitioningFactoryOptions();
  options.infer_dictionary = infer_dictionary;
  auto factory = HivePartitioning::MakeFactory(options);

  ASSIGN_OR_ABORT(auto schema, factory->Inspect({path}));
  ASSIGN_OR_ABORT(auto partitioning, factory->Finish(schema));
  ASSIGN_OR_ABORT(auto expr, partitioning->Parse(path));
  return expr;
}
```

This way the dict and non-dict partition expressions are derived equivalently